clang/test/Frontend/sycl-propagate-aspect-warning.cpp

```diff
@@ -0,0 +1,97 @@
+// RUN: %clang_cc1 -internal-isystem %S/Inputs -fsycl-is-device -mconstructor-aliases -no-opaque-pointers -emit-codegen-only -verify %s
+
+// Tests for warnings when propagated aspects do not match the aspects available
+// in a function, as specified through the 'sycl::device_has' attribute.
+
+#include "sycl.hpp"
+
+// expected-note@+1 4 {{propagated from call to function 'Struct1::Struct1()'}}
+struct [[__sycl_detail__::__uses_aspects__(sycl::aspect::fp16)]] Struct1 {
+  int a = 0;
+};
+
+// expected-note@+1 4 {{propagated from call to function 'func3(int, int, int)'}}
+int func3(int a, int b, int c) {
+  Struct1 s;
+  s.a = 1;
+  return s.a;
+}
+
+// expected-note@+1 4 {{propagated from call to function 'func4(int, int)'}}
+int func4(int a, int b) {
+  double x = 3.0;
+  return func3(a, b, (int)x);
+}
+
+// expected-note@+1 2 {{propagated from call to function 'func2(int, int)'}}
+int func2(int a, int b) { return func3(a, b, 1); }
+
+// expected-warning@+1 {{function 'func1(int)' uses aspect 'fp16' not listed in its 'sycl::device_has' attribute}}
+[[sycl::device_has(sycl::aspect::fp64)]] int func1(int a) { return func2(a, 1); }
+
+int main() {
+  sycl::queue Q;
+  Q.submit([&](sycl::handler &CGH) {
+    CGH.single_task([=]() { int a = func1(1); });
+  });
+  Q.submit([&](sycl::handler &CGH) {
+    // expected-warning-re@+1 {{function '{{.*}}' uses aspect 'fp16' not listed in its 'sycl::device_has' attribute}}
```

Comment:
Is this warning accurate? device_has is not specified on the function. It is specified on the kernel from which the function is called right? Also can the diagnostic be emitted at call site? This will prevent the need to include function name in the diagnostic.

Reply:
>> I intentionally filter out the kernel itself, as otherwise it will give both a warning and a note for the same lambdas.

Can you please explain why? Aren't warnings being emitted on first caller and notes on remaining? Or am I misunderstanding something?

>> I think we could, but it would mean we would have to attach srcloc to all callsites inside functions with the device_has attribute, making this approach even more aggressive. If you think it is worth it, I am open to trying it.

Hmm...I'm not sure. @premanandrao can you please weigh in? We're not emitting the warning on the problematic call but rather on the caller (I think). This isn't ideal but to emit warnings on actual callsite, we'll have to attach source location metadata to all call sites so that the information is available to the pass which emits these diagnostics. I recall an earlier conversation where there were concerns about potential increase in binary size and compilation time in current approach. I assume this more aggressive approach will amplify those concerns.

Reply:
> Can you please explain why? Aren't warnings being emitted on first caller and notes on remaining? Or am I misunderstanding something?

We seem to have two layers to the kernel entry points, whereas the outer layer is carrying the `device_has` attribute. Since both layers have the same source location we would get the warning from the outer function, then a note about it calling the inner function.

Maybe I misunderstood your original question. Did you mean we should change the warning to something like "kernel function uses ..." when the warning is from a kernel? That should be possible too, though I suppose it also depends on whether we want to move the diagnostics to the callsites.

@premanandrao - Friendly ping on the question above.

Reply:
@elizabethandrews and @steffenlarsen, sorry for the delay in getting back on this.

> We're not emitting the warning on the problematic call but rather on the caller (I think). This isn't ideal but to emit warnings on actual callsite, we'll have to attach source location metadata to all call sites so that the information is available to the pass which emits these diagnostics.

I looked at the diagnostic generated and while it is different from how we generally identify the issue, I did not find it confusing.  To me, it seemed clear what it was pointing out, and the subsequent notes pointed out where the violation stemmed from.

Based on that, I wouldn't advocate the extra cost of attaching source location info to all the call sites.

Perhaps this could be done later if we get user feedback that the message is unhelpful?

Reply:
Yes, in this case it would be the kernel name. it is a little verbose because of its mangling, but if you think it's better I can make the test explicit.

Reply:
I don't think its required.